### ISSUE TITLE
An Idea About Improving Documentation

### ISSUE THREAD

**Issue by walkr** [phoenixframework/phoenix#1452]
Programming is a fun but complex activity, especially in this day and age, when people work with multiple languages, juggle with various mental models – both of the language and of the program/system – and contribute and extend other people's code.

I like Phoenix and Elixir a lot, but I sometimes find the documentation hard to follow, and I'm assuming a lot of other people might be in the same situation, especially if they are introduced to a new language – in this case Elixir – through a new system/program (Phoenix). 

I think we would all benefit if the documentation would achieve the goal of making it easier for the new programmer to not just understand what a component does, but also why the whole system was designed a certain way.

I actually just finished reading an interesting [piece] (http://pages.cs.wisc.edu/~remzi/Naur.pdf) about programming, which prompted me to write this. This piece proposes an interesting view of what programming is:

> [...] programming should be regarded as an activity by which the programmers form or achieve a certain kind of insight, a theory of the matters at hand.

I found particularly insightful the idea of writing documentation which will help the next programmer build an adequate theory of the program. Describing the theory of the program which inspired the design is more important than simply describing the design.

Other interesting points about writing documentation:

1. Creating metaphors which describe the system
2. Writing text describing the purpose of each major component
3. Drawing schematics for the major interactions between the major components

I'm actually surprised how little we make use of schematics in documenting code nowadays.

Here is a link to the text in case anyone is interested to read the whole piece: [Peter Naur – Programming as Theory Building [pdf]] (http://pages.cs.wisc.edu/~remzi/Naur.pdf)

Cheers!

**Comment by chrismccord** [comment]
Hey Tony,

Thanks for the ideas and feedback! We try to make the docs as approachable and comprehensive as possible, but we leave some expansion on design to the guides. 

> sometimes find the documentation hard to follow

Please raise an issue in such cases and we'll do our best to fill in the gaps.

> I'm actually surprised how little we make use of schematics in documenting code nowadays.

I agree these are nice, and this one is probably more a time bottleneck than anything. You can find open issues on the guides that are for schematics of different interactions.

In the future, feel free to raise an issue for docs that are confusing. We also gladly accept doc and guide PR for review. Thanks!
